File: python/src/agent_api/local/files.py
from __future__ import annotations

import base64
import errno
import shutil
import time
from pathlib import Path
from typing import Any, Literal

from .errors import LocalFileTooLargeError, LocalNotTextFileError, LocalPathError
from .paths import (
    assert_inside_root,
    atomic_write,
    entry_from_stat,
    file_type,
    ignored,
    is_likely_text_file,
    looks_binary,
    mime_type,
    normalize_relative_path,
    patch_line_range,
    portable_path,
    positive_int,
    select_line_range,
    split_lines,
)
from .types import LocalFileStat, LocalIgnoreRule


class LocalFileStore:
    def __init__(self, root: str | Path, *, label: str = "local") -> None:
        self.root = Path(root).expanduser().resolve()
        self.label = label

    def child(self, relative_path: str | Path, *, label: str | None = None) -> "LocalFileStore":
        return LocalFileStore(self.resolve_path(relative_path), label=label or self.label)

    def ensure(self) -> None:
        self.root.mkdir(parents=True, exist_ok=True)

    def resolve_path(self, relative_path: str | Path = ".") -> Path:
        clean = normalize_relative_path(relative_path)
        full_path = (self.root / clean).resolve()
        assert_inside_root(self.root, full_path)
        return full_path

    def relative_path(self, full_path: str | Path) -> str:
        absolute = Path(full_path).expanduser().resolve()
        assert_inside_root(self.root, absolute)
        return portable_path(absolute.relative_to(self.root))

    def exists(self, relative_path: str | Path = ".") -> bool:
        return self.resolve_path(relative_path).exists()

    def stat(self, relative_path: str | Path = ".") -> LocalFileStat:
        full_path = self.resolve_path(relative_path)
        info = full_path.lstat()
        return LocalFileStat(
            path=portable_path(full_path.relative_to(self.root)) or ".",
            full_path=str(full_path),
            type=file_type(full_path, info),
            size=info.st_size,
            modified_at=info.st_mtime,
        )

    def mkdir(self, relative_path: str | Path = ".") -> Path:
        full_path = self.resolve_path(relative_path)
        full_path.mkdir(parents=True, exist_ok=True)
        return full_path

    def list(
        self,
        relative_path: str | Path = ".",
        *,
        recursive: bool = False,
        include_directories: bool = False,
        max_depth: int | None = None,
        ignore: list[LocalIgnoreRule] | None = None,
    ) -> list[LocalFileStat]:
        base = self.resolve_path(relative_path)
        depth_limit = max_depth if recursive else 1
        out: list[LocalFileStat] = []
        self._walk(self.root, base, base, out, depth_limit, include_directories, ignore or [])
        return sorted(out, key=lambda item: item.path)

    def list_entries(self, relative_path: str | Path = ".", **options: Any) -> dict[str, Any]:
        options.setdefault("include_directories", True)
        return {"object": "list", "entries": [entry_from_stat(item) for item in self.list(relative_path, **options)]}

    def search_entries(self, *, query: str, path: str | Path = ".", limit: int = 100) -> dict[str, Any]:
        needle = query.strip().lower()
        if not needle:
            raise ValueError("query is required")
        entries = [
            entry_from_stat(item)
            for item in self.list(path, recursive=True, include_directories=True)
            if needle in item.path.lower()
        ][: positive_int(limit, 100)]
        return {"object": "list", "entries": entries}

    def read_text(self, relative_path: str | Path) -> str:
        return self.resolve_path(relative_path).read_text(encoding="utf-8")

    def read_json(self, relative_path: str | Path) -> Any:
        import json

        return json.loads(self.read_text(relative_path))

    def read_bytes(self, relative_path: str | Path) -> bytes:
        return self.resolve_path(relative_path).read_bytes()

    def read_file(self, relative_path: str | Path, *, max_bytes: int | None = None, format: Literal["text", "raw"] = "text") -> dict[str, Any]:
        full_path = self.resolve_path(relative_path)
        if not full_path.is_file():
            raise LocalPathError("local path is not a file", str(relative_path))
        raw = full_path.read_bytes()
        limit = max_bytes if max_bytes is not None else len(raw)
        truncated = len(raw) > limit
        content = raw[:limit] if truncated else raw
        portable = portable_path(full_path.relative_to(self.root))
        if format == "raw":
            return {"path": portable, "size": len(raw), "truncated": truncated, "content": content, "content_type": mime_type(portable)}
        if looks_binary(content):
            return {
                "path": portable,
                "encoding": "base64",
                "mime_type": mime_type(portable),
                "size": len(raw),
                "truncated": truncated,
                "content_base64": base64.b64encode(content).decode("ascii"),
            }
        return {
            "path": portable,
            "encoding": "text",
            "mime_type": mime_type(portable),
            "size": len(raw),
            "truncated": truncated,
            "content": content.decode("utf-8"),
        }

    def write_text(self, relative_path: str | Path, content: str, *, atomic: bool = True) -> Path:
        full_path = self.resolve_path(relative_path)
        full_path.parent.mkdir(parents=True, exist_ok=True)
        if atomic:
            atomic_write(full_path, content.encode("utf-8"))
        else:
            full_path.write_text(content, encoding="utf-8")
        return full_path

    def write_json(self, relative_path: str | Path, value: Any, *, pretty: bool = True, atomic: bool = True) -> Path:
        import json

        indent = 2 if pretty else None
        return self.write_text(relative_path, json.dumps(value, indent=indent) + "\n", atomic=atomic)

    def write_bytes(self, relative_path: str | Path, content: bytes, *, atomic: bool = True) -> Path:
        full_path = self.resolve_path(relative_path)
        full_path.parent.mkdir(parents=True, exist_ok=True)
        if atomic:
            atomic_write(full_path, content)
        else:
            full_path.write_bytes(content)
        return full_path

    def write_file(self, relative_path: str | Path, content: str | bytes, *, atomic: bool = True) -> dict[str, Any]:
        full_path = self.write_text(relative_path, content, atomic=atomic) if isinstance(content, str) else self.write_bytes(relative_path, content, atomic=atomic)
        return {"path": portable_path(full_path.relative_to(self.root)), "size": full_path.stat().st_size}

    def remove(self, relative_path: str | Path) -> None:
        full_path = self.resolve_path(relative_path)
        if full_path.is_dir():
            shutil.rmtree(full_path, ignore_errors=True)
        else:
            full_path.unlink(missing_ok=True)

    def delete_path(self, relative_path: str | Path) -> dict[str, Any]:
        self.remove(relative_path)
        return {"path": normalize_relative_path(relative_path), "recursive": True}

    def create_directory(self, relative_path: str | Path = ".") -> dict[str, Any]:
        self.mkdir(relative_path)
        return {"path": normalize_relative_path(relative_path)}

    def copy(self, from_relative_path: str | Path, to_relative_path: str | Path) -> Path:
        source = self.resolve_path(from_relative_path)
        target = self.resolve_path(to_relative_path)
        target.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(source, target)
        return target

    def read_lines(self, relative_path: str | Path, *, start_line: int, end_line: int | None = None, max_bytes: int | None = None) -> dict[str, Any]:
        file = self.read_file(relative_path, max_bytes=max_bytes, format="raw")
        if looks_binary(file["content"]):
            raise LocalNotTextFileError(str(relative_path))
        text = file["content"].decode("utf-8")
        lines = split_lines(text)
        selected, resolved_end = select_line_range(lines, start_line, end_line)
        return {
            "path": file["path"],
            "start_line": int(start_line),
            "end_line": resolved_end,
            "total_lines": len(lines),
            "lines": selected,
            "file_truncated": file["truncated"],
            "size": file["size"],
        }

    def patch_lines(
        self,
        relative_path: str | Path,
        *,
        start_line: int,
        end_line: int | None = None,
        replacement: str = "",
        max_bytes: int | None = None,
    ) -> dict[str, Any]:
        file = self.read_file(relative_path, max_bytes=max_bytes, format="raw")
        if file["truncated"]:
            raise LocalFileTooLargeError(str(relative_path))
        if looks_binary(file["content"]):
            raise LocalNotTextFileError(str(relative_path))
        patched, resolved_end, total = patch_line_range(file["content"].decode("utf-8"), start_line, end_line, replacement)
        written = self.write_file(relative_path, patched)
        return {"path": file["path"], "start_line": int(start_line), "end_line": resolved_end, "total_lines": total, "size": written["size"]}

    def grep(
        self,
        *,
        pattern: str,
        path: str | Path = ".",
        limit: int = 200,
        max_files: int = 500,
        max_bytes_per_file: int = 512 * 1024,
        max_line_length: int = 500,
        ignore: list[LocalIgnoreRule] | None = None,
    ) -> dict[str, Any]:
        needle = pattern.strip()
        if not needle:
            raise ValueError("pattern is required")
        matches: list[dict[str, Any]] = []
        files_scanned = 0
        scan_truncated = False
        for item in self._grep_candidates(path, ignore or []):
            if len(matches) >= limit or files_scanned >= max_files:
                scan_truncated = True
                break
            if item.type != "file" or item.size > max_bytes_per_file or not is_likely_text_file(item.path):
                continue
            raw = read_optional_bytes(Path(item.full_path))
            if raw is None:
                continue
            if looks_binary(raw):
                continue
            files_scanned += 1
            for idx, line in enumerate(split_lines(raw.decode("utf-8")), start=1):
                if needle not in line:
                    continue
                matches.append({"path": item.path, "line_number": idx, "line": line[:max_line_length]})
                if len(matches) >= limit:
                    scan_truncated = True
                    break
        return {"object": "list", "matches": matches, "files_scanned": files_scanned, "scan_truncated": scan_truncated}

    def _grep_candidates(self, relative_path: str | Path, ignore: list[LocalIgnoreRule]) -> list[LocalFileStat]:
        full_path = self.resolve_path(relative_path)
        if not full_path.exists():
            raise FileNotFoundError(errno.ENOENT, "No such file or directory", str(full_path))
        portable = portable_path(full_path.relative_to(self.root)) or "."
        if ignored(portable, ignore):
            return []
        info = full_path.lstat()
        if full_path.is_file():
            return [LocalFileStat(portable, str(full_path), file_type(full_path, info), info.st_size, info.st_mtime)]
        if full_path.is_dir():
            return self.list(relative_path, recursive=True, ignore=ignore)
        return []

    def summarize(
        self,
        *,
        path: str | Path = ".",
        max_files: int = 2000,
        max_previews: int = 20,
        preview_bytes: int = 4096,
        top_paths: int = 20,
        max_depth: int | None = None,
        ignore: list[LocalIgnoreRule] | None = None,
    ) -> dict[str, Any]:
        stats = self.list(path, recursive=True, max_depth=max_depth, ignore=ignore)
        all_files = [item for item in stats if item.type == "file"]
        files = all_files[:max_files]
        by_size = sorted(files, key=lambda item: (-item.size, item.path))
        previews: list[dict[str, Any]] = []
        for item in by_size:
            if len(previews) >= max_previews:
                break
            if not is_likely_text_file(item.path) or item.size > preview_bytes * 4:
                continue
            raw = read_optional_bytes(Path(item.full_path))
            if raw is None:
                continue
            if looks_binary(raw):
                continue
            previews.append(
                {
                    "path": item.path,
                    "size": item.size,
                    "preview": raw[:preview_bytes].decode("utf-8"),
                    "preview_truncated": len(raw) > preview_bytes or None,
                }
            )
        return {
            "summary_path": "",
            "file_count": len(files),
            "total_bytes": sum(item.size for item in files),
            "top_paths_by_size": [f"{item.path} ({item.size} bytes)" for item in by_size[:top_paths]],
            "text_previews": previews,
            "generated_at_unix": int(time.time()),
            "scan_truncated": len(all_files) > len(files),
        }

    def _walk(
        self,
        store_root: Path,
        scan_root: Path,
        directory: Path,
        out: list[LocalFileStat],
        max_depth: int | None,
        include_directories: bool,
        ignore: list[LocalIgnoreRule],
    ) -> None:
        if not directory.exists():
            return
        for entry in sorted(directory.iterdir(), key=lambda item: item.name):
            rel = portable_path(entry.absolute().relative_to(store_root))
            if ignored(rel, ignore):
                continue
            info = lstat_optional(entry)
            if info is None:
                continue
            stat = LocalFileStat(rel, str(entry), file_type(entry, info), info.st_size, info.st_mtime)
            if entry.is_dir() and not entry.is_symlink():
                if include_directories:
                    out.append(stat)
                depth = len(portable_path(entry.absolute().relative_to(scan_root)).split("/"))
                if max_depth is None or depth < max_depth:
                    self._walk(store_root, scan_root, entry, out, max_depth, include_directories, ignore)
            else:
                out.append(stat)


def lstat_optional(path: Path) -> Any | None:
    try:
        return path.lstat()
    except OSError as exc:
        if exc.errno in {errno.ENOENT, errno.ENOTDIR}:
            return None
        raise


def read_optional_bytes(path: Path) -> bytes | None:
    try:
        return path.read_bytes()
    except OSError as exc:
        if exc.errno in {errno.ENOENT, errno.ENOTDIR, errno.EISDIR}:
            return None
        raise

File: go/agentapi/local/files.go
package local

import (
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"syscall"
	"time"
	"unicode/utf8"
)

const maxScanWarnings = 20

type FileStore struct {
	Root  string
	Label string
}

type ListOptions struct {
	Recursive          bool
	IncludeDirectories bool
	MaxDepth           int
	Ignore             []IgnoreRule
}

type ReadFileParams struct {
	MaxBytes int
}

type ReadLinesParams struct {
	StartLine int
	EndLine   int
	MaxBytes  int
}

type PatchLinesParams struct {
	StartLine   int
	EndLine     int
	Replacement string
	MaxBytes    int
}

type GrepParams struct {
	Pattern         string
	Path            string
	Limit           int
	MaxFiles        int
	MaxBytesPerFile int
	MaxLineLength   int
	Ignore          []IgnoreRule
}

type SummaryParams struct {
	Path         string
	MaxFiles     int
	MaxPreviews  int
	PreviewBytes int
	TopPaths     int
	MaxDepth     int
	Ignore       []IgnoreRule
}

func NewFileStore(root string, label string) (*FileStore, error) {
	abs, err := filepath.Abs(root)
	if err != nil {
		return nil, err
	}
	if label == "" {
		label = "local"
	}
	return &FileStore{Root: abs, Label: label}, nil
}

func (s *FileStore) Child(rel string, label string) (*FileStore, error) {
	full, _, err := ResolveInside(s.Root, rel)
	if err != nil {
		return nil, err
	}
	if label == "" {
		label = s.Label
	}
	return NewFileStore(full, label)
}

func (s *FileStore) Ensure() error {
	return os.MkdirAll(s.Root, 0o755)
}

func (s *FileStore) ResolvePath(rel string) (string, error) {
	full, _, err := ResolveInside(s.Root, rel)
	return full, err
}

func (s *FileStore) RelativePath(fullPath string) (string, error) {
	abs, err := filepath.Abs(fullPath)
	if err != nil {
		return "", err
	}
	root, _ := filepath.Abs(s.Root)
	rel, err := filepath.Rel(root, abs)
	if err != nil {
		return "", err
	}
	if strings.HasPrefix(rel, ".."+string(filepath.Separator)) || rel == ".." || filepath.IsAbs(rel) {
		return "", pathError("local path must stay inside the store root", fullPath)
	}
	return filepath.ToSlash(rel), nil
}

func (s *FileStore) Exists(rel string) bool {
	full, err := s.ResolvePath(rel)
	if err != nil {
		return false
	}
	_, err = os.Stat(full)
	return err == nil
}

func (s *FileStore) Stat(rel string) (*FileStat, error) {
	full, portable, err := ResolveInside(s.Root, rel)
	if err != nil {
		return nil, err
	}
	info, err := os.Lstat(full)
	if err != nil {
		return nil, err
	}
	return &FileStat{Path: firstNonEmpty(portable, "."), FullPath: full, Type: fileType(info), Size: info.Size(), ModifiedAt: info.ModTime()}, nil
}

func (s *FileStore) Mkdir(rel string) (string, error) {
	full, err := s.ResolvePath(rel)
	if err != nil {
		return "", err
	}
	return full, os.MkdirAll(full, 0o755)
}

func (s *FileStore) List(rel string, opts ListOptions) ([]FileStat, error) {
	stats, _, err := s.listWithWarnings(rel, opts)
	return stats, err
}

func (s *FileStore) listWithWarnings(rel string, opts ListOptions) ([]FileStat, []ScanWarning, error) {
	base, _, err := ResolveInside(s.Root, rel)
	if err != nil {
		return nil, nil, err
	}
	maxDepth := opts.MaxDepth
	if !opts.Recursive {
		maxDepth = 1
	}
	var out []FileStat
	var warnings []ScanWarning
	err = s.walk(s.Root, base, base, maxDepth, opts, &out, &warnings)
	sort.Slice(out, func(i, j int) bool { return out[i].Path < out[j].Path })
	return out, warnings, err
}

func (s *FileStore) ListEntries(rel string, opts ListOptions) (*EntryList, error) {
	opts.IncludeDirectories = true
	stats, warnings, err := s.listWithWarnings(rel, opts)
	if err != nil {
		return nil, err
	}
	entries := make([]Entry, 0, len(stats))
	for _, item := range stats {
		entries = append(entries, entryFromStat(item))
	}
	return &EntryList{Object: "list", Entries: entries, ScanWarnings: warnings}, nil
}

func (s *FileStore) SearchEntries(query, rel string, limit int) (*EntryList, error) {
	query = strings.ToLower(strings.TrimSpace(query))
	if query == "" {
		return nil, fmt.Errorf("query is required")
	}
	if limit <= 0 {
		limit = 100
	}
	stats, err := s.List(rel, ListOptions{Recursive: true, IncludeDirectories: true})
	if err != nil {
		return nil, err
	}
	var entries []Entry
	for _, item := range stats {
		if strings.Contains(strings.ToLower(item.Path), query) {
			entries = append(entries, entryFromStat(item))
			if len(entries) >= limit {
				break
			}
		}
	}
	return &EntryList{Object: "list", Entries: entries}, nil
}

func (s *FileStore) ReadText(rel string) (string, error) {
	raw, err := s.ReadBytes(rel)
	return string(raw), err
}

func (s *FileStore) ReadJSON(rel string, out any) error {
	raw, err := s.ReadBytes(rel)
	if err != nil {
		return err
	}
	return json.Unmarshal(raw, out)
}

func (s *FileStore) ReadBytes(rel string) ([]byte, error) {
	full, err := s.ResolvePath(rel)
	if err != nil {
		return nil, err
	}
	return os.ReadFile(full)
}

func (s *FileStore) ReadFile(rel string, params ReadFileParams) (*FileDeliver, error) {
	raw, info, portable, err := s.readRaw(rel, params.MaxBytes)
	if err != nil {
		return nil, err
	}
	truncated := params.MaxBytes > 0 && int64(len(raw)) < info.Size()
	deliver := &FileDeliver{Path: portable, MimeType: mimeType(portable), Size: info.Size(), Truncated: truncated}
	if looksBinary(raw) || !utf8.Valid(raw) {
		deliver.Encoding = "base64"
		deliver.ContentBase64 = base64.StdEncoding.EncodeToString(raw)
	} else {
		deliver.Encoding = "text"
		deliver.Content = string(raw)
	}
	return deliver, nil
}

func (s *FileStore) ReadFileRaw(rel string, params ReadFileParams) (*FileRaw, error) {
	raw, info, portable, err := s.readRaw(rel, params.MaxBytes)
	if err != nil {
		return nil, err
	}
	return &FileRaw{Path: portable, Size: info.Size(), Truncated: params.MaxBytes > 0 && int64(len(raw)) < info.Size(), Content: raw, ContentType: mimeType(portable)}, nil
}

func (s *FileStore) WriteText(rel, content string) (string, error) {
	return s.WriteTextAtomic(rel, content, true)
}

func (s *FileStore) WriteTextAtomic(rel, content string, atomic bool) (string, error) {
	full, err := s.ResolvePath(rel)
	if err != nil {
		return "", err
	}
	if err := os.MkdirAll(filepath.Dir(full), 0o755); err != nil {
		return "", err
	}
	if atomic {
		return full, atomicWrite(full, []byte(content))
	}
	return full, os.WriteFile(full, []byte(content), 0o644)
}

func (s *FileStore) WriteJSON(rel string, value any) (string, error) {
	raw, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		return "", err
	}
	return s.WriteBytes(rel, append(raw, '\n'))
}

func (s *FileStore) WriteBytes(rel string, content []byte) (string, error) {
	return s.WriteBytesAtomic(rel, content, true)
}

func (s *FileStore) WriteBytesAtomic(rel string, content []byte, atomic bool) (string, error) {
	full, err := s.ResolvePath(rel)
	if err != nil {
		return "", err
	}
	if err := os.MkdirAll(filepath.Dir(full), 0o755); err != nil {
		return "", err
	}
	if atomic {
		return full, atomicWrite(full, content)
	}
	return full, os.WriteFile(full, content, 0o644)
}

func (s *FileStore) WriteFile(rel string, content []byte) (*FileWrite, error) {
	full, err := s.WriteBytes(rel, content)
	if err != nil {
		return nil, err
	}
	info, err := os.Stat(full)
	if err != nil {
		return nil, err
	}
	portable, _ := s.RelativePath(full)
	return &FileWrite{Path: portable, Size: info.Size()}, nil
}

func (s *FileStore) Remove(rel string) error {
	full, err := s.ResolvePath(rel)
	if err != nil {
		return err
	}
	return os.RemoveAll(full)
}

func (s *FileStore) DeletePath(rel string) (*PathDelete, error) {
	clean, err := NormalizeRelativePath(rel)
	if err != nil {
		return nil, err
	}
	if err := s.Remove(rel); err != nil {
		return nil, err
	}
	return &PathDelete{Path: clean, Recursive: true}, nil
}

func (s *FileStore) CreateDirectory(rel string) (map[string]string, error) {
	clean, err := NormalizeRelativePath(rel)
	if err != nil {
		return nil, err
	}
	_, err = s.Mkdir(clean)
	return map[string]string{"path": clean}, err
}

func (s *FileStore) Copy(fromRel, toRel string) (string, error) {
	from, err := s.ResolvePath(fromRel)
	if err != nil {
		return "", err
	}
	to, err := s.ResolvePath(toRel)
	if err != nil {
		return "", err
	}
	raw, err := os.ReadFile(from)
	if err != nil {
		return "", err
	}
	if err := os.MkdirAll(filepath.Dir(to), 0o755); err != nil {
		return "", err
	}
	return to, os.WriteFile(to, raw, 0o644)
}

func (s *FileStore) ReadLines(rel string, params ReadLinesParams) (*FileLines, error) {
	file, err := s.ReadFileRaw(rel, ReadFileParams{MaxBytes: params.MaxBytes})
	if err != nil {
		return nil, err
	}
	if looksBinary(file.Content) || !utf8.Valid(file.Content) {
		return nil, notTextFileError(rel)
	}
	lines := splitLines(string(file.Content))
	selected, end, err := selectLineRange(lines, params.StartLine, params.EndLine)
	if err != nil {
		return nil, err
	}
	return &FileLines{Path: file.Path, StartLine: params.StartLine, EndLine: end, TotalLines: len(lines), Lines: selected, FileTruncated: file.Truncated, Size: file.Size}, nil
}

func (s *FileStore) PatchLines(rel string, params PatchLinesParams) (*FileLinesPatch, error) {
	file, err := s.ReadFileRaw(rel, ReadFileParams{MaxBytes: params.MaxBytes})
	if err != nil {
		return nil, err
	}
	if file.Truncated {
		return nil, fileTooLargeError(rel)
	}
	if looksBinary(file.Content) || !utf8.Valid(file.Content) {
		return nil, notTextFileError(rel)
	}
	patched, end, total, err := patchLineRange(string(file.Content), params.StartLine, params.EndLine, params.Replacement)
	if err != nil {
		return nil, err
	}
	written, err := s.WriteFile(rel, []byte(patched))
	if err != nil {
		return nil, err
	}
	return &FileLinesPatch{Path: file.Path, StartLine: params.StartLine, EndLine: end, TotalLines: total, Size: written.Size}, nil
}

func (s *FileStore) Grep(params GrepParams) (*GrepResponse, error) {
	pattern := strings.TrimSpace(params.Pattern)
	if pattern == "" {
		return nil, fmt.Errorf("pattern is required")
	}
	path := firstNonEmpty(params.Path, ".")
	limit := firstPositiveInt(params.Limit, 200)
	maxFiles := firstPositiveInt(params.MaxFiles, 500)
	maxBytes := firstPositiveInt(params.MaxBytesPerFile, 512*1024)
	maxLine := firstPositiveInt(params.MaxLineLength, 500)
	stats, err := s.grepCandidates(path, params.Ignore)
	if err != nil {
		return nil, err
	}
	resp := &GrepResponse{Object: "list"}
	for _, item := range stats {
		if len(resp.Matches) >= limit || resp.FilesScanned >= maxFiles {
			resp.ScanTruncated = true
			break
		}
		if item.Type != FileTypeFile || item.Size > int64(maxBytes) || !isLikelyTextFile(item.Path) {
			continue
		}
		raw, err := os.ReadFile(item.FullPath)
		if err != nil || looksBinary(raw) || !utf8.Valid(raw) {
			continue
		}
		resp.FilesScanned++
		for i, line := range splitLines(string(raw)) {
			if !strings.Contains(line, pattern) {
				continue
			}
			if len(line) > maxLine {
				line = line[:maxLine]
			}
			resp.Matches = append(resp.Matches, GrepMatch{Path: item.Path, LineNumber: i + 1, Line: line})
			if len(resp.Matches) >= limit {
				resp.ScanTruncated = true
				break
			}
		}
	}
	return resp, nil
}

func (s *FileStore) grepCandidates(rel string, ignore []IgnoreRule) ([]FileStat, error) {
	full, portable, err := ResolveInside(s.Root, rel)
	if err != nil {
		return nil, err
	}
	info, err := os.Lstat(full)
	if err != nil {
		return nil, err
	}
	if portable == "" {
		portable = "."
	}
	if Ignored(portable, ignore) {
		return nil, nil
	}
	if info.IsDir() {
		return s.List(rel, ListOptions{Recursive: true, Ignore: ignore})
	}
	if info.Mode().IsRegular() {
		return []FileStat{{
			Path:       portable,
			FullPath:   full,
			Type:       FileTypeFile,
			Size:       info.Size(),
			ModifiedAt: info.ModTime(),
		}}, nil
	}
	return nil, nil
}

func (s *FileStore) Summarize(params SummaryParams) (*Summary, error) {
	path := firstNonEmpty(params.Path, ".")
	maxFiles := firstPositiveInt(params.MaxFiles, 2000)
	maxPreviews := firstPositiveInt(params.MaxPreviews, 20)
	previewBytes := firstPositiveInt(params.PreviewBytes, 4096)
	topPaths := firstPositiveInt(params.TopPaths, 20)
	stats, warnings, err := s.listWithWarnings(path, ListOptions{Recursive: true, MaxDepth: params.MaxDepth, Ignore: params.Ignore})
	if err != nil {
		return nil, err
	}
	var files []FileStat
	for _, item := range stats {
		if item.Type == FileTypeFile {
			files = append(files, item)
		}
	}
	scanTruncated := len(files) > maxFiles
	if len(files) > maxFiles {
		files = files[:maxFiles]
	}
	bySize := append([]FileStat(nil), files...)
	sort.Slice(bySize, func(i, j int) bool {
		if bySize[i].Size != bySize[j].Size {
			return bySize[i].Size > bySize[j].Size
		}
		return bySize[i].Path < bySize[j].Path
	})
	out := &Summary{SummaryPath: "", FileCount: len(files), GeneratedAtUnix: time.Now().Unix(), ScanTruncated: scanTruncated, ScanWarnings: warnings}
	for _, item := range files {
		out.TotalBytes += item.Size
	}
	for i, item := range bySize {
		if i >= topPaths {
			break
		}
		out.TopPathsBySize = append(out.TopPathsBySize, fmt.Sprintf("%s (%d bytes)", item.Path, item.Size))
	}
	for _, item := range bySize {
		if len(out.TextPreviews) >= maxPreviews {
			break
		}
		if !isLikelyTextFile(item.Path) || item.Size > int64(previewBytes*4) {
			continue
		}
		raw, err := os.ReadFile(item.FullPath)
		if err != nil || looksBinary(raw) || !utf8.Valid(raw) {
			continue
		}
		truncated := len(raw) > previewBytes
		if truncated {
			raw = raw[:previewBytes]
		}
		out.TextPreviews = append(out.TextPreviews, SummaryPreview{Path: item.Path, Size: item.Size, Preview: string(raw), PreviewTruncated: truncated})
	}
	return out, nil
}

func (s *FileStore) readRaw(rel string, maxBytes int) ([]byte, os.FileInfo, string, error) {
	full, portable, err := ResolveInside(s.Root, rel)
	if err != nil {
		return nil, nil, "", err
	}
	info, err := os.Stat(full)
	if err != nil {
		return nil, nil, "", err
	}
	if !info.Mode().IsRegular() {
		return nil, nil, "", pathError("local path is not a file", rel)
	}
	raw, err := os.ReadFile(full)
	if err != nil {
		return nil, nil, "", err
	}
	if maxBytes > 0 && len(raw) > maxBytes {
		raw = raw[:maxBytes]
	}
	return raw, info, portable, nil
}

func (s *FileStore) walk(storeRoot, scanRoot, dir string, maxDepth int, opts ListOptions, out *[]FileStat, warnings *[]ScanWarning) error {
	entries, err := os.ReadDir(dir)
	if err != nil {
		if dir != scanRoot && recordScanWarning(warnings, portableRel(storeRoot, dir), err) {
			return nil
		}
		return err
	}
	for _, entry := range entries {
		full := filepath.Join(dir, entry.Name())
		rel, err := filepath.Rel(storeRoot, full)
		if err != nil {
			return err
		}
		portable := filepath.ToSlash(rel)
		if Ignored(portable, opts.Ignore) {
			continue
		}
		info, err := os.Lstat(full)
		if err != nil {
			if recordScanWarning(warnings, portable, err) {
				continue
			}
			return err
		}
		stat := FileStat{Path: portable, FullPath: full, Type: fileType(info), Size: info.Size(), ModifiedAt: info.ModTime()}
		if info.IsDir() {
			if opts.IncludeDirectories {
				*out = append(*out, stat)
			}
			depthRel, _ := filepath.Rel(scanRoot, full)
			depth := 1
			if depthRel != "." {
				depth = len(strings.Split(filepath.ToSlash(depthRel), "/"))
			}
			if maxDepth <= 0 || depth < maxDepth {
				if err := s.walk(storeRoot, scanRoot, full, maxDepth, opts, out, warnings); err != nil {
					return err
				}
			}
			continue
		}
		*out = append(*out, stat)
	}
	return nil
}

func fileType(info os.FileInfo) FileType {
	if info.Mode()&os.ModeSymlink != 0 {
		return FileTypeSymlink
	}
	if info.IsDir() {
		return FileTypeDirectory
	}
	if info.Mode().IsRegular() {
		return FileTypeFile
	}
	return FileTypeOther
}

func entryFromStat(item FileStat) Entry {
	size := item.Size
	if item.Type == FileTypeDirectory {
		size = 0
	}
	return Entry{Path: item.Path, IsDir: item.Type == FileTypeDirectory, Size: size, ModifiedAtUnix: item.ModifiedAt.Unix()}
}

func portableRel(root, full string) string {
	rel, err := filepath.Rel(root, full)
	if err != nil || rel == "." {
		return "."
	}
	return filepath.ToSlash(rel)
}

func recordScanWarning(warnings *[]ScanWarning, portable string, err error) bool {
	if !isIgnorableScanError(err) {
		return false
	}
	if len(*warnings) < maxScanWarnings {
		*warnings = append(*warnings, ScanWarning{
			Path:    firstNonEmpty(portable, "."),
			Code:    scanErrorCode(err),
			Message: err.Error(),
		})
	}
	return true
}

func isIgnorableScanError(err error) bool {
	return errors.Is(err, os.ErrNotExist) ||
		errors.Is(err, os.ErrPermission) ||
		errors.Is(err, syscall.ENOTDIR) ||
		errors.Is(err, syscall.ENOTCONN) ||
		errors.Is(err, syscall.ELOOP) ||
		errors.Is(err, syscall.EINVAL)
}

func scanErrorCode(err error) string {
	for _, candidate := range []struct {
		err  error
		code string
	}{
		{os.ErrNotExist, "ENOENT"},
		{os.ErrPermission, "EACCES"},
		{syscall.ENOTDIR, "ENOTDIR"},
		{syscall.ENOTCONN, "ENOTCONN"},
		{syscall.ELOOP, "ELOOP"},
		{syscall.EINVAL, "EINVAL"},
	} {
		if errors.Is(err, candidate.err) {
			return candidate.code
		}
	}
	return ""
}
